### PR TITLE
More search improvements

### PR DESCRIPTION
**Pull request by JJ-Pineda**
Resolves [#106 ](https://github.com/rdkit-rs/cheminee/issues/106) by adding more multithreading to the underlying structure search methods so that fingerprint screening and structure matches can be parallelized.

There is also some IO improvement: there were some places where we were opening the same docs multiple times to access the same data we were working with earlier. Now the docs are only opened once and the relevant data for the passing compounds are recorded as we go along.

### CODE REVIEW

File: src/search/identity_search.rs [hunk 1]
```diff
@@ -38,22 +39,29 @@ pub fn identity_search(
...
         if fp_match {
             let mol_exact_match =
                 exact_match(&ROMol::from_smiles(smiles)?, query_mol, use_chirality);
             if mol_exact_match {
-                filtered_results.insert(docaddr);
+                let extra_data = match doc.get_first(extra_data_field) {
+                    Some(extra_data) => serde_json::to_string(
+                        extra_data
+                            .as_json()
+                            .ok_or(eyre::eyre!("Failed to jsonify extra data"))?,
```
Comment (xrl) [inline src/search/identity_search.rs]: 👏

File: tests/search_tests.rs [hunk 2]
```diff
@@ -48,9 +48,13 @@ fn test_identity_search() {
     let smiles_field = builder.add_text_field("smiles", STRING | STORED);
     let fingerprint_field = builder.add_bytes_field("fingerprint", FAST | STORED);
 
+    let json_options: JsonObjectOptions =
+        JsonObjectOptions::from(TEXT | STORED).set_expand_dots_enabled();
+    let _extra_data_field = builder.add_json_field("extra_data", json_options);
```
Comment (xrl) [inline tests/search_tests.rs:53]: I would just delete this line if the variable isn't used.

**Review by xrl — CHANGES_REQUESTED** [review]

File: tests/search_tests.rs (hunk 2 above)
Comment (JJ-Pineda) [reply tests/search_tests.rs:53]: So the builder is still adding the extra data field, which gets incorporated into the schema with this line. I tried deleting it and it blew up at me:)

Comment (xrl) [reply tests/search_tests.rs:53]: Ah, my bad. Just don't assign it to a variable, remove everything to the left of the word `builder`.

**Review by xrl — APPROVED** [review]

File: tests/search_tests.rs (hunk 2 above)
Comment (JJ-Pineda) [reply tests/search_tests.rs:53]: Sounds good. I figured clippy might complain if I didn't assign a dummy variable, but maybe not?

**Pull request merged**